### PR TITLE
ui: Distinguish bot users in more places

### PR DESCRIPTION
**Pull request by sm-sayedi**
Now that #605 is merged already, bot icons are shown with the name of bot users in `MessageListPage`. Additional places where users appear are covered in this PR:

- `ProfilePage`
- `RecentDmConversationsPage`

### ProfilePage
<table>
    <tr>
        <td><b>Before</b></td>
        <td><b>After</b></td>
        <td><b>After (with long name)</b></td>
    </tr>
    <tr>
        <td></td>
        <td></td>
        <td></td>
    </tr>
</table>

### RecentDmConversationsPage
<table>
    <tr>
        <td><b>Before</b></td>
        <td><b>After</b></td>
        <td><b>After (with long name)</b></td>
    </tr>
    <tr>
        <td></td>
        <td></td>
        <td></td>
    </tr>
</table>

~**Note:** In `RecentDmConversationsPage` I wasn't sure whether to add (and where to add) the bot icon to a group dm where there may be one or multiple bot recipients.~

Fixes: #636

### CODE REVIEW

**Review by rajveermalviya — APPROVED** [review]
Thanks! Changes LGTM and tests well :)
Moving on to mentor review, cc @hackerkid.

File: lib/widgets/recent_dm_conversations.dart
Comment (hackerkid) [inline lib/widgets/recent_dm_conversations.dart:120]: Why is this always false? Isn't it possible to create DMs with multiple bots?

Comment (sm-sayedi) [reply lib/widgets/recent_dm_conversations.dart:120]: > In `RecentDmConversationsPage`, I wasn't sure whether to add (and where to add) the bot icon to a group dm where there may be one or multiple bot recipients.

As I mentioned in the PR description above, I wasn't quite sure about this. There are two cases in group DMs with bot users: 

- A group DM with both real and bot users; in which I think it is not reasonable to add a bot icon at the end of the label as we do it in `1:1` DM with bot user. One approach we can take is to add a bot icon after the name of each bot user, something like this: **Welcome Bot 🤖, Mahmood, S Muhmood**. Or we can ignore the bot icon at all. I am not sure.
- A group DM with just bot users; in which I think we can add a bot icon at the end of the label, but again, it will be confused with the first case, is it that the group DM is with multiple real users and one last bot user or the group DM is with all bot users?! If we ignore the bot icon in the first case, then this cannot be confusing!

Let's also hear from @gnprice about this!

Comment (gnprice) [reply lib/widgets/recent_dm_conversations.dart:120]: This is a question about how the UI should behave for users, so it's a good one to get feedback on from @alya .

Comment (alya) [reply lib/widgets/recent_dm_conversations.dart:120]: The bot icon should be associated with a particular user, not with a conversation as a whole. So I would expect it to appear after each bot's name, regardless of how many bot and non-bot users there are in the conversation.

Comment (alya) [reply lib/widgets/recent_dm_conversations.dart:120]: By the way, https://github.com/zulip/zulip/issues/26831 is the corresponding web app issue.

**Comment by gnprice** [comment]
> The bot icon should be associated with a particular user, not with a conversation as a whole. So I would expect it to appear after each bot's name

@sm-sayedi For how to implement this behavior, the key Flutter feature I think you'll want is [WidgetSpan](https://main-api.flutter.dev/flutter/widgets/WidgetSpan-class.html). For example in message content that's how we handle custom emoji, user mentions, and global times.

File: lib/widgets/recent_dm_conversations.dart
Comment (hackerkid) [reply lib/widgets/recent_dm_conversations.dart:120]: That makes a lot of sense. Thanks for the clarification @alya!

**Comment by sm-sayedi** [comment]
@hackerkid, @alya Revision pushed with images also updated in the PR description. PTAL.

**Review by hackerkid — APPROVED** [review]

**Comment by alya** [comment]
We would ideally show the icon even if the name gets abbreviated.

**Comment by gnprice** [comment]
Advancing to maintainer review, since mentor review is complete.

In general reviewers should advance the PR to the next stage when they're done reviewing (or done except a few small comments). It's also a good idea for PR authors to check in to see if that hasn't happened, and advance it themselves.

For cross-reference, there's a chat thread started from Alya's comment above:
https://chat.zulip.org/#narrow/stream/48-mobile/topic/bot.20icon.20in.20recent-dms.20items.20-.20.23F764/near/1846941

That doesn't affect the first commit in this PR, though. Not sure if it'll affect the later commits enough to hold off reviewing them until it's been revised for that.

File: lib/widgets/recent_dm_conversations.dart
Comment (chrisbobbe) [inline lib/widgets/recent_dm_conversations.dart:160]: I see a comment on the `switch`:

```dart
 // TODO dedupe with DM items in [InboxPage]
```

and that reminds me that we also show users' names in the Inbox page, when there are unread DMs. Would you add a prep commit that makes a helper function that can be used in this page and the Inbox page too? Perhaps it can return a `TextSpan`, and it can live in either file or perhaps a new file.

File: test/widgets/recent_dm_conversations_test.dart
Comment (chrisbobbe) [inline test/widgets/recent_dm_conversations_test.dart:266]:
```dart
        group('no error when user somehow missing from store.users', () {
          // […]
          testWidgets('bot recipient -> shows no bot icon', (tester) async {
            await checkRecipient(tester, isBot: true, looksBot: false);
          });

          testWidgets('non-bot recipient -> shows no bot icon', (tester) async {
            await checkRecipient(tester, isBot: false, looksBot: false);
          });
```

The "bot recipient" test doesn't actually simulate a bot recipient, right, because these tests are checking what happens when the user data is missing. I see that the `User` passed to `eg.dmMessage` has `isBot: true`, but the `isBot` doesn't actually get written into the `DmMessage` object, so the app code can't treat this case differently than the "non-bot recipient" case.

Simplest I think to leave this test is it was, but with an added `checkBotIcon(count: 0);`.

File: test/widgets/recent_dm_conversations_test.dart
Comment (chrisbobbe) [inline test/widgets/recent_dm_conversations_test.dart:304]: We would like the bot icon to appear even when the name is long; see Alya's comment: https://github.com/zulip/zulip-flutter/pull/764#issuecomment-2207386553

If that's hard to do and has to be left as a TODO, that might be OK, but we shouldn't have a test that confirms the undesired behavior. 🙂

File: lib/widgets/profile.dart
Comment (chrisbobbe) [inline lib/widgets/profile.dart:66]: I think it would look better to style this like we do in the other places: first the name, then the bot icon, as an inline `WidgetSpan`. Is the style in this revision based on something you'd like me to look at?

File: lib/widgets/profile.dart
Comment (chrisbobbe) [inline lib/widgets/profile.dart:57]: Let's put this in the new `DesignVariables`, and use the new variable here and in the message list (where we added the bot icon for senders in #605).

**Review by chrisbobbe — COMMENTED** [review]
Thanks @sm-sayedi! Comments below.

**Comment by gnprice** [comment]
Thanks @sm-sayedi for your work on this, and @hackerkid @chrisbobbe for the reviews!

After reviewing our priorities over the past week, this issue #636 is one I'm content to leave open past the launch of the app. So let's leave this PR for it aside; it'll serve as a basis to start from when someone comes back later, after launch, to pick up the issue and finish it. I'll close the PR to get it out of our dashboards since we won't actively be working on it, but it'll remain linked from the issue so easy to find when the time comes.

In the meantime, we have plenty of [other issues](https://github.com/orgs/zulip/projects/5/views/1) to focus our energy on :slightly_smiling_face:

**Pull request closed without merging**
